File: impl/csharp/test/StreamingBoundsTests.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System.Collections.Generic;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// The stream chunk-count bound (design.md §3.4, R7), ported from
    /// impl/go/streaming/bounds_test.go. A commit over exactly MaxStreamChunks chunks verifies, and
    /// one over MaxStreamChunks+1 is rejected TooManyChunks. Both carry a MATCHING rolling digest, so
    /// the count is the only reason to reject — deleting the count check makes the +1 case verify
    /// (the mutation is caught). The chunks share one backing array to bound test memory.
    /// </summary>
    public sealed class StreamingBoundsTests
    {
        [Fact]
        public void BoundTooManyChunks()
        {
            var over = new List<Streaming.Chunk>((int)Envelope.MaxStreamChunks + 1);
            for (int i = 0; i <= Envelope.MaxStreamChunks; i++)
            {
                over.Add(new Streaming.Chunk(0, new byte[0]));
            }
            var atLimit = over.GetRange(0, (int)Envelope.MaxStreamChunks);

            var okCommit = new Streaming.StreamCommit(new byte[0], Streaming.CommitDigest(atLimit));
            Streaming.VerifyCommit(okCommit, atLimit); // must not throw

            var overCommit = new Streaming.StreamCommit(new byte[0], Streaming.CommitDigest(over));
            var ex = Assert.Throws<NaalpException>(() => Streaming.VerifyCommit(overCommit, over));
            Assert.Equal("TooManyChunks", ex.Kind);

            // VerifyCheckpoint enforces the same bound, and the count check fires before the
            // contiguity/digest checks, so the diagnosis is TooManyChunks (not a digest error).
            var cpEx = Assert.Throws<NaalpException>(() =>
                Streaming.VerifyCheckpoint(new Streaming.StreamCheckpoint(new byte[0], 0, new byte[0]), over));
            Assert.Equal("TooManyChunks", cpEx.Kind);
        }
    }
}

File: impl/kotlin/src/test/kotlin/StreamingBoundsTests.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

/**
 * Stream chunk-count bound (design.md §3.4, R7) known-answer test for the Kotlin SDK, mirroring
 * impl/go/streaming/bounds_test.go: a commit over exactly MAX_STREAM_CHUNKS chunks verifies, and
 * one over MAX_STREAM_CHUNKS+1 is rejected TooManyChunks. Both carry a MATCHING rolling digest, so
 * the count is the only reason to reject -- deleting the count check makes the +1 case verify (the
 * mutation is caught). The chunks share offset 0 with empty data (a preallocated cheap accept-at-
 * limit case, mirroring Go's zero-value chunk slice) so building MAX_STREAM_CHUNKS+1 of them stays
 * fast.
 *
 * Run (main()-driven, no test framework):
 *   kotlinc src/main/kotlin src/test/kotlin/StreamingBoundsTests.kt -cp <bcprov.jar> -include-runtime -d out/streaming-bounds.jar
 *   java -cp "out/streaming-bounds.jar;<bcprov.jar>" sh.bubblefish.naalp.StreamingBoundsTestsKt
 */

private class StreamingBoundsFailure(msg: String) : RuntimeException(msg)

private fun sbOk(what: String) = println("  ok  $what")

private val SB_EMPTY = ByteArray(0)

/** n zero-offset, empty-data chunks: trivially contiguous (each satisfies offset==runningTotal==0),
 *  so the same list serves both the commit and the checkpoint boundary checks. */
private fun sbZeroChunks(n: Int): List<Streaming.Chunk> = List(n) { Streaming.Chunk(0L, SB_EMPTY) }

/** Run [block]; return "ok" if it returns, else the NaalpException kind. */
private fun sbVerifyKind(block: () -> Unit): String = try {
    block()
    "ok"
} catch (e: NaalpException) {
    e.kind
}

// TestBoundTooManyChunks pins the stream chunk-count bound (design.md §3.4, R7): a commit over
// exactly MAX_STREAM_CHUNKS chunks verifies, and one over MAX_STREAM_CHUNKS+1 is rejected
// TooManyChunks.
private fun testBoundTooManyChunks() {
    val overCount = (MAX_STREAM_CHUNKS + 1L).toInt()
    val over = sbZeroChunks(overCount)
    val atLimit = over.subList(0, MAX_STREAM_CHUNKS.toInt())

    val okCommit = Streaming.StreamCommit(SB_EMPTY, Streaming.commitDigest(atLimit))
    val okKind = sbVerifyKind { Streaming.verifyCommit(okCommit, atLimit) }
    if (okKind != "ok") throw StreamingBoundsFailure("commit over $MAX_STREAM_CHUNKS chunks should verify, got $okKind")
    sbOk("commit at MaxStreamChunks verifies")

    val overCommit = Streaming.StreamCommit(SB_EMPTY, Streaming.commitDigest(over))
    val overKind = sbVerifyKind { Streaming.verifyCommit(overCommit, over) }
    if (overKind != "TooManyChunks") {
        throw StreamingBoundsFailure("commit over ${MAX_STREAM_CHUNKS + 1L} chunks should be TooManyChunks, got $overKind")
    }
    sbOk("commit over MaxStreamChunks rejected TooManyChunks")

    // VerifyCheckpoint enforces the same bound, and the count check fires before the
    // contiguity/digest checks, so the diagnosis is TooManyChunks (not a digest error).
    val cpKind = sbVerifyKind {
        Streaming.verifyCheckpoint(Streaming.StreamCheckpoint(SB_EMPTY, 0L, SB_EMPTY), over)
    }
    if (cpKind != "TooManyChunks") {
        throw StreamingBoundsFailure("checkpoint over ${MAX_STREAM_CHUNKS + 1L} chunks should be TooManyChunks, got $cpKind")
    }
    sbOk("checkpoint over MaxStreamChunks rejected TooManyChunks (count checked before digest)")
}

fun main() {
    println("StreamingBoundsTests -- stream chunk-count bound (design.md §3.4, R7)")
    testBoundTooManyChunks()
    println("StreamingBoundsTests: PASS")
}
